### DISCUSSION TITLE
Resuming a fuzz session after corpus trimming

### DISCUSSION THREAD

**Discussion by alexanderceleste** [AFLplusplus/AFLplusplus#1254] [Q&A]
Hello,

First, a quick thanks to you folks. AFL++ is an awesome tool which I've greatly enjoyed using.

Now, my current target, for a variety of reasons, tends to generate false positives. To avoid saturating my corpus with useless inputs, I have been quitting the fuzz session intermittently and using afl-cmin to identify and remove useless inputs. More specifically, I will fuzz for one hour, trim useless inputs to reduce corpus size by 25-40%, and then fuzz again using this trimmed corpus.

However, my issue is that AFL++ does not like resuming a fuzzing session after inputs are removed. If I start with 100 inputs and then stop fuzzing to use afl-cmin and trim down to 75, when I resume the fuzz session it will think that I still have 100 inputs and crash/throw an error when it tries to access an input that has been deleted. To work around this, I have been starting each new fuzzing session "from scratch" by moving the trimmed corpus into in/. 

Is there a better way to handle this? I know AFL++ keeps a lot of state while fuzzing, and I am losing that state by starting "from scratch" each time. Could I write a script to make resuming possible, perhaps by editing certain files which track the number of corpus inputs (e.g. plot_data, fuzzer_stats, etc.)?

Or, is a solution not necessary because starting "from scratch" each time is a fine workaround?

Thanks in advance, 
AC

**Comment by vanhauser-thc** [comment]
there is not really "state" that is recovered when you resume. there is no difference between resuming and starting a new fuzzing instance except you keep the stats (time it was running so far, the plot etc.).

also there is limited value in stoping, using afl-cmin and then restarting.

also having some unnecessary inputs over time is not a big issue. afl randomly selects the entries to fuzz and gives those more interesting (more edges, more unique edges, etc.) a higher chance to be selected, so uninteresting ones are not so often touched.

I recommend rather to find out why there are so many unnecessary inputs. high instability of edges in your target? lots of threading?

(and sorry, a queue with just 100 entries sounds rather empty to me, that is not even a good starting corpus ;)

**Reply by alexanderceleste** [reply]
I really appreciate the quick response! Thank you.

My fuzz target is a closed source binary instrumented via DynamoRIO. The false positives are partially due to multithreading, and possibly also some quirks with my DynamoRIO setup causing instability as well. I have spent some time trying to fully diagnose the issue but haven't been successful yet, unfortunately. I agree that fixing the instability is the core issue here.

Also, my pitiful queue size is likely owed to my fuzzing speed - I am at only 25-30 execs/sec, even with a deferred forkserver. However, I have been fuzzing using 5 instances over the span of a couple of weeks... so, over 100 million execs. Maybe my target simply does not have that many paths.

It seems like leaving the fuzzer running and ignoring false positives could be a better approach, but I worry that fuzzer effectiveness might diminish. You said having "some" unnecessary inputs over time is not a big issue, but what if over half of the corpus contains false positives after maybe 2-4 hours (200k-400k execs)? Is that too much, and worth trimming?

**Reply by vanhauser-thc** [reply]
oh I am curious - how are you fuzzing the target with dynamorio? We do not provide anything for that in afl++ :)
but very likely you will get much better results using qemu_mode or frida_mode (speed wise and features)

**Reply by alexanderceleste** [reply]
I wrote a small DynamoRIO plugin specifically to handle the deferred fork server and code coverage reporting. It works pretty well - I tell AFL++ to fuzz drrun, which is a DynamoRIO binary that runs my target with my plugin. The plugin is a library that performs the initial handshake with AFL, then instruments all basic blocks with a few assembly instructions which report code coverage to AFL via the SHM. Then, when the program runs, DynamoRIO monitors it and waits for the fork server address, and starts the fork server once it gets there.

I looked at qemu_mode before, but it looks like it needs LLVM in order to have a deferred fork server, which is a challenge for me.

frida_mode might work, but I can't tell whether I can still have deferred fork server with that approach? There's a lot of initialization I need to skip so the deferred fork server is important :)

**Reply by vanhauser-thc** [reply]
qemu_mode, nyx_mode and frida_mode are all for binary only targets and hence do not need any llvm. qemu_mode and frida_mode both can put the forkserver wherever you want, both also support persistent mode if your target work with that.
but best of all it will be ~100+ times faster than your dynamorio solution.

**Reply by vanhauser-thc** [reply]
you should read this document thoroughly: https://github.com/AFLplusplus/AFLplusplus/blob/dev/docs/fuzzing_binary-only_targets.md

**Reply by alexanderceleste** [reply]
At this point I can't even recall why I decided not to try any of those in the beginning. I will say I have a very challenging target, so maybe I thought a custom solution would be best, but at this point it is worth trying new approaches.

Thanks for the help!

**Reply by alexanderceleste** [reply]
Hello again,

I spent some time looking into QEMU and FRIDA mode. Your docs say FRIDA is often slightly faster than QEMU, and FRIDA appears to be a dynamic binary instrumentation tool like DynamoRIO. Specifically, it looks like AFL uses FRIDA to insert a deferred forkserver into the code, and to insert assembly snippets into basic blocks to report code coverage to AFL. After those two tasks, fuzzing begins. Is that roughly correct? 

If yes, that is also what my DynamoRIO plugin does - it instruments basic blocks once upfront with assembly snippets, then loads them into the DynamoRIO code cache to execute more quickly ("near-natively" according to DynamoRIO) once fuzzing begins. Since both frida_mode and my DynamoRIO plugin do the same thing, I am not sure I understand how frida_mode could be over 100 times faster... unless FRIDA itself is 100x faster than DynamoRIO (which doesn't seem likely!)

I apologize for asking so much about this, but both qemu_mode and frida_mode will be difficult for me to build and setup for my target, so I want to be sure it will be worthwhile to explore these before I dedicate a lot of my time to it. Thanks again for your guidance.

**Reply by domenukk** [reply]
DynamoRIO doesn't seem to be on par with FRIDA when it comes to speed, but maybe your instrumentation is better optimized. It could make sense to try your instrumentation against frida-mode on a toy binary to see the performance difference.
Also, there are further features like cmplog support that you very likely will want to have (solving compares during fuzzing)

**Reply by vanhauser-thc** [reply]
I cant say it better than dominik. try it. then you know. 100x was maybe a bit high, but I expect a 10x, and if you can use persistent mode then you easily get x20-x50. plus the cmplog feature.
